### PR TITLE
Upgrade Rust to 1.52.0

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

Rust announced 1.52.0

#### Summary of Changes

Update solana repo to Rust 1.52.0

Fixes #
Taking following commands to upgrade Rust: 
```$ rustup upgrade; rustup --version 
$ cargo +nightly check --all --tests --benches```

### CODE REVIEW

**Comment by taozhu-chicago** [comment]
run `rustup toolchain install 1.52.0` on my local box seem to fix below failed check: 
```cargo +1.52.0 clippy --version --verbose
error: toolchain '1.52.0-x86_64-apple-darwin' is not installed```
@mvines is this something to be done on build box, or can be part of PR?

File: ci/rust-version.sh [hunk 1]
```diff
@@ -18,7 +18,7 @@
 if [[ -n $RUST_STABLE_VERSION ]]; then
   stable_version="$RUST_STABLE_VERSION"
 else
-  stable_version=1.51.0
+  stable_version=1.52.0
```
[the remark this answers is not in the document]
Comment (mvines) [reply ci/rust-version.sh:21]: also update the nightly below to:
```
nightly_version=2021-05-10
```

**Comment by taozhu-chicago** [comment]
> This new error might require some consulting with @ryoqun, the AbiExample expert
> 
> ```
> error: cannot specialize on trait `Default`
> --
>   | --> version/src/lib.rs:11:59
>   | \|
>   | 11 \| #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, AbiExample)]
> ```

Thanks @mvines, @ryoqun if there is anything to try, please feel free let me know. Thanks!

**Comment by ryoqun** [comment]
> Thanks @mvines, @ryoqun if there is anything to try, please feel free let me know. Thanks!

@taozhu-chicago hi, I think I've fixed the frozen_abi compilation issue. it's hard to debug if you're not familiar with the code because of the heavy macro magic...  `cargo expand` could be useful in that case... :)

**Comment by taozhu-chicago** [comment]
> > Thanks @mvines, @ryoqun if there is anything to try, please feel free let me know. Thanks!
> 
> @taozhu-chicago hi, I think I've fixed the frozen_abi compilation issue. it's hard to debug if you're not familiar with the code because of the heavy macro magic... `cargo expand` could be useful in that case... :)

@ryoqun thanks for fixing it ... and the `expand` hint! I'll take it from here to continue updating codebase for 1.52.0

**Comment by taozhu-chicago** [comment]
what do you guys think about this new error: 
```
error: avoid using `collect()` when not needed
  | --> runtime/src/accounts.rs:846:14
  | \|
  | 846 \|             .collect();
  | \|              ^^^^^^^
  | 847 \|         let mut account_locks = &mut self.account_locks.lock().unwrap();
  | 848 \|         keys.into_iter()
  | \|         ---------------- the iterator could be used here instead
  | \|
  | = note: `-D clippy::needless-collect` implied by `-D warnings`
  | = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
  | help: use the original Iterator instead of collecting it and then producing a new one
  | \|
  | 834 \|
  | 835 \|         let mut account_locks = &mut self.account_locks.lock().unwrap();
  | 836 \|         txs
  | 837 \|             .map(\|tx\| {
  | 838 \|                 tx.sanitize().map_err(TransactionError::from)?;
  | 839 \|
  | ...
 
```
Is it OK to add `#[allow(clippy::needless_collect)]` to keep the existing code's intention of minimize account lock window?

**Comment by mvines** [comment]
> Is it OK to add `#[allow(clippy::needless_collect)]` to keep the existing code's intention of minimize account lock window?

It looks to me that it's likely a push. The checks that are done before acquiring the lock seem minor: 
https://github.com/solana-labs/solana/blob/794700e81b82d6f5cdf05d7804f9a3959bc4ed00/runtime/src/accounts.rs#L828-L857
Would love a second 👍🏼 on that though

**Comment by taozhu-chicago** [comment]
> > Is it OK to add `#[allow(clippy::needless_collect)]` to keep the existing code's intention of minimize account lock window?
> 
> It looks to me that it's likely a push. The checks that are done before acquiring the lock seem minor:
> https://github.com/solana-labs/solana/blob/794700e81b82d6f5cdf05d7804f9a3959bc4ed00/runtime/src/accounts.rs#L828-L857
> 
> 
> Would love a second 👍🏼 on that though

Since the sanity/dup check is read operation, it makes sense to do that before acquiring lock.  @carllin  @jeffwashington your thoughts?

**Comment by jeffwashington** [comment]
> Is it OK to add `#[allow(clippy::needless_collect)]` to keep the existing code's intention of minimize account lock window?

`#[allow(clippy::needless_collect)]` is my recommendation. I spend time finding places where we are holding locks too long and fixing the situations ;-) Definitely if your goal is upgrading rust, then I would not change locking behavior intentionally in the same pr. If you want to rework the lock in a separate pr, we could discuss just that issue there.

Seems like an over aggressive clippy rule to me. The rule is ignoring side effects. You could consider filing a rust/clippy issue with example code like this.

**Comment by taozhu-chicago** [comment]
to address this test coverage error: 
```
/usr/local/cargo/bin/grcov
--
  | Error: Required grcov version not installed
  | Installed version: grcov 0.8.0
  | 🚨 Error: The command exited with status 1

```
I updated `scripts/coverage.sh` to include grcov version `0.8.0`. To highlight and seek 👍 s for this change.

**Comment by mvines** [comment]
> to address this test coverage error:
> 
> ```
> /usr/local/cargo/bin/grcov
> --
>   | Error: Required grcov version not installed
>   | Installed version: grcov 0.8.0
>   | 🚨 Error: The command exited with status 1
> ```
> 
> I updated `scripts/coverage.sh` to include grcov version `0.8.0`. To highlight and seek 👍 s for this change.

yep that's fine.  We have to bump the grcov version sometimes

**Comment by taozhu-chicago** [comment]
create issue https://github.com/solana-labs/solana/issues/17228 to properly address `bare-trait-objects` warning, I allowed this warning to continue upgrading.

**Review by mvines — APPROVED** [review]
lgtm once CI passes!

**Comment by mvines** [comment]
So close!  All green, only to be dinged with a merge conflict.

**Comment by CriesofCarrots** [comment]
> So close! All green, only to be dinged with a merge conflict.

😳 mea culpa, so sorry!

**Pull request merged**